### PR TITLE
[CALCITE-922] Extract the value of an INTERVAL constant properly in t…

### PR DESCRIPTION
**Pull request by hsuanyi**
…he case where this constant times a CAST() function

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/sql/SqlLiteral.java [hunk 1]
```diff
@@ -284,6 +284,11 @@ public static Comparable value(SqlNode node) {
         return (NlsString) literal.value;
       case NUMERIC:
         return (BigDecimal) literal.value;
+      case INTERVAL_YEAR_MONTH:
+      case INTERVAL_DAY_TIME:
+        final SqlIntervalLiteral.IntervalValue val =
+            (SqlIntervalLiteral.IntervalValue) literal.value;
+        return val.getSign();
```
Comment (vlsi) [inline core/src/main/java/org/apache/calcite/sql/SqlLiteral.java]: Why is that?
I think `getSign` is not appropriate here

Comment (hsuanyi) [inline core/src/main/java/org/apache/calcite/sql/SqlLiteral.java]: The code path we are observing is: (think of it as stack trace)
SqlLiteral.value() (line: 291)
SqlCallBinding.getOperandLiteralValue() (line: 142)
SqlMonotonicBinaryOperator.getMonotonicity() (line: 77)

So for an expression such as  

cast(...) * (INTERVAL '-1' DAY) 

To make the monotonicity correct, we need to let SqlLiteral.value() (line: 291) return -1. 

Make sense?

Comment (hsuanyi) [inline core/src/main/java/org/apache/calcite/sql/SqlLiteral.java]: You could try the added unit test to reproduce the path.

Comment (vlsi) [inline core/src/main/java/org/apache/calcite/sql/SqlLiteral.java]: >To make the monotonicity correct, we need to let SqlLiteral.value() (line: 291) return -1.
>Make sense?

It does not.
You are changing `value` method. It's javadoc says `* Extracts the value from a literal.`.
It's javadoc explicitly says  `If the node is a {@link SqlIntervalQualifier...` and you are returning just sign.

That does not work.

**Pull request closed without merging**